ncm-systemd/src/main/perl/Systemd/UnitFile.pm

```diff
@@ -0,0 +1,444 @@
+        if (! ($self->_directory_exists($unitdir) || $self->_make_directory($unitdir))) {
+            $self->error("Failed to create unitdir $unitdir: $!");
+            return;
+        }
+    };
+
+    return $filename;
+}
+
+=item _hwloc_calc_cpuaffinity
+
+Run C<_hwloc_calc_cpus>, and returns in C<CPUAffinity> format with a reset
+
+=cut
+
+sub _hwloc_calc_cpuaffinity
+{
+    my ($self, $locations) = @_;
+
+    my $cpus = $self->_hwloc_calc_cpus($locations);
+    return if(! defined($cpus));
+
+    # first empty list, to reset all previous defined CPUaffinity settings
+    return [[], $cpus];
+}
+
+
+=item _hwloc_calc_cpus
+
+Run the C<hwloc-calc --physical --intersect PU> command for C<locations>.
+
+Returns arrayref with CPU indices on success, undef otherwise.
+
+=cut
+
+sub _hwloc_calc_cpus
+{
+    my ($self, $locations) = @_;
+
+    local $ENV{HWLOC_HIDE_ERRORS};
+    $ENV{HWLOC_HIDE_ERRORS}=1;
+
+    # pass a copy of the Readonly array, so we can extend it
+    my $proc = CAF::Process->new([@HWLOC_CALC_CPUS], log => $self);
+    $proc->pushargs(@$locations);
+
+    my $output = $proc->output();
+    my @indices = grep { $_ =~ m/^\d+$/ } split(/,/, $output);
+    if (join(',', @indices) eq $output) {
+        return \@indices;
+    } else {
+        $self->error("Unexpected output from $proc: $output");
+        return;
+    }
+}
+
+
+=item _make_variables_custom
+
+A function that return the custom variables hashref to pass as ttoptions.
+(This is a function, not a method).
+
+=cut
+
+sub _make_variables_custom {
+    my $customs = shift;
+    my $ttoptions;
+    $ttoptions->{VARIABLES}->{SYSTEMD}->{CUSTOM} = $customs;
+    return $ttoptions;
+}
+
+# TODO: Move to CAF::AllTheMissingBitsThatLCProvides
+
+# make directory, mkdir -p style, simple mkpath wrapper
+sub _make_directory
+{
+    my ($self, $directory) = @_;
+    return mkpath($directory);
+}
+
+# -d, wrapped in method for unittesting
+# -d follows symlink, a broken symlink either exists with -l or not
+# and can be cleaned up with rmtree
+sub _directory_exists
+{
+    my ($self, $directory) = @_;
+    return (! -l $directory) && -d $directory;
+}
+
+# -f, wrapped in method for unittesting
+sub _file_exists
+{
+    my ($self, $filename) = @_;
+    return (-f $filename || -l $filename);
+}
+
+# exists, -e || -l, wrapped in method for unittesting
+# LC::Check::_unlink uses lstat and -e _ (is that a single FS query?)
+sub _exists
+{
+    my ($self, $path) = @_;
+    return -e $path || -l $path;
+}
+
+# _unlink, remove with backup support
+# works like LC::Check::_unlink, but has directory support
+# and no error throwing
+# returns SUCCESS on success, undef on failure, logs error
+# backup is backup from LC::Chekc::_unlink (and thus also CAF::File*)
+# if backup is undefined, use self->{backup}
+# pass empty string to disable backup with self->{backup} defined
+# does not cleanup the backup of the original file,
+# FileWriter via TextRender can do that.
+sub _cleanup
```

Comment:
Are callers expected to do the NoAction check? I think it would be preferable if the low level function did it to avoid programmer mistakes.